Doc update: How to capture group events (#6214)
PostHog/posthog.com: contents/docs/product-analytics/group-analytics.mdx

<JSCreateGroups />

## How to set group properties

<JSSetGroupProperties />

## How to capture group events

> Events are automatically linked to the groups that the user who submitted the event belongs to. Thus you do not add any additional code once you have linked a user to the group. This is the recommended way to capturing group events.

Reviewer: ```suggestion
In the JavaScript web library, groups are automatically linked to the user who triggers an event. Thus you do not add any additional code once you have linked a user to a group. This is the recommended way to capturing group events.
```
This is only true on the frontend?

Reviewer: From what I can tell, this isn't the case. See this example:
![Screen Shot 2023-07-07 at 8 29 28 AM](https://github.com/PostHog/posthog.com/assets/34755028/ced40502-b0ae-4669-8ab4-736b992e2095)

user_signed_up event happened first with groups properties, then finished onboarding, but finished onboarding doesn't have group properties.

Reviewer 2: Nope, this isn't true, ever.

There's no concept of a user belonging to a group, so events for a person are never linked to a group. You always have to mention groups property on an event to tell us which group this belongs to.

Just like how every event requires a distinct ID to tell us which user it belongs to.

The difference/confusion here might be this:
1. On the client side, we "save" the group, if any, and add it to all events automatically.
2. On the server side, you explicitly have to add group to all events.

This is similar to how you don't need to add a distinct ID on the client side because we save it, but you do on the server side, since it might not be the same user for the subsequent requests.

Author: Thanks both for calling this out!

> You always have to mention groups property on an event to tell us which group this belongs to.

I'll update the docs to reflect this.

> To link a user to a group, call the same code as you would when [creating a group](#how-to-create-groups).

Reviewer: ```suggestion
To link a user to a group using the web library, call the same code as you would when [creating a group](#how-to-create-groups).
```

### Advanced (server-side only): Capturing group events without a user

If you want to capture group events but don't want to associate them with a specific user, there are two alternatives:

#### Method one (recommended): Use the group ID as the event's distinct ID

When capturing events, instead of using the user's ID as the distinct ID, you can use group ID as the distinct ID. For example:

<MultiLanguage>

```node
posthog.capture({
    event: 'group_event_name',
    distinctId: 'group_id_in_your_db',
})
```

```python
posthog.capture('group_id_in_your_db', 
    'group_event_name', 
)
```

```go
client.Enqueue(posthog.Capture{
    DistinctId: "group_id_in_your_db",
    Event: "group_event_name",
})
```

```php
PostHog::capture(array(
    'distinctId' => 'group_id_in_your_db',
    'event' => 'group_event_name'
));
```

```bash
curl -v -L --header "Content-Type: application/json" -d '{
    "api_key": "<ph_project_api_key>",
    "event": "group_event_name",
    "distinct_id": "group_id_in_your_db",
}' <ph_instance_address>/capture/

> ```

Reviewer: Add `groups` properties so these play nicely with group analytics?

> This has the advantage that insights such as "total users and "unique users" in [trends](/docs/product-analytics/trends) will work as expected for these groups events (note: although it will say "users", it will actually mean "groups")

Reviewer: This assumes you are only using total users and unique users for groups and not individual users. If you do have individual users too, it gets messed up.

Author: This is per event though right? 
So depending on the event, "total users" would mean users or groups
e.g. for pageviews it would be users, but for `group_event` it would be group.

Perhaps I should clarify this in the text

Reviewer: Fair point, there is also the `All events` series  to worry about.

Reviewer 2: Since there's no link between users and groups, you shouldn't ever need to link these two together, even in insights.

If you want total count of events -> just do total count.
If you want unique count per group -> just do unique group.

user doesn't exist anyway, so no need to do total users.

If users do have meaning, then you should have used the distinct ID of the user in the first place? 🤔 . Am I missing something here?

> The drawback of this approach is that a new [person](/docs/data/persons) will be created for each group. While this won't affect your data integrity in any way, it may cause confusion since these groups are not actually people.

Reviewer: See above comment, this does affect data integrity if you also have normal "persons"

</MultiLanguage>

#### Method two: Use the same ID as the event's distinct ID

> Instead of using the group's ID as the distinct ID, you can create a *single string* that will be used to capture group events. This can be anything you want, as long as it's the same string for every group event:

Reviewer: ```suggestion
Instead of using the group's ID as the distinct ID, you can use a *single value* to capture group events. This can be anything you want, as long as it's the same for every group event:
```

```node
posthog.capture({
    event: 'group_event_name',
    distinctId: 'static_string_for_group_events',
})
```

```python
posthog.capture('static_string_for_group_events', 
    'group_event_name', 
)
```

```go
client.Enqueue(posthog.Capture{
    DistinctId: "static_string_for_group_events",
    Event: "group_event_name",
})
```

```php
PostHog::capture(array(
    'distinctId' => 'static_string_for_group_events',
    'event' => 'group_event_name'
));
```

```bash
curl -v -L --header "Content-Type: application/json" -d '{
    "api_key": "<ph_project_api_key>",
    "event": "group_event_name",
    "distinct_id": "static_string_for_group_events",
}' <ph_instance_address>/capture/

> ```

Reviewer: Should add `groups` properties and a note about varying them between groups (like on normal users).

Author: Hmm Im actually not sure this approach is possible. Its not possible to assign the same user to multiple groups of the same type.

@tiina303 any thoughts?

Reviewer: On the backend, yes. Groups are set on events, not persons on the backend.

![Screen Shot 2023-07-07 at 8 36 41 AM](https://github.com/PostHog/posthog.com/assets/34755028/6efcf1ab-0791-4f01-ab92-b86dcb73bdef)

> The drawback of this approach is that insights such as "total users and "unique users" in [trends](/docs/product-analytics/trends) will not work as expected for these group events (since it is same ID submitting these events).

Reviewer: If you just use one ID for group analytics, total users and unique users will be more accurate for normal persons (1 more unique user for groups)

Author: When in posthog would you calculate or use "total users" or "unique users" for ALL normal persons? 

These are done usually at an event level. Im not aware of other use cases?

Reviewer: There is an `All events` series.

Author: Ah, didnt realize we had this! Thanks for point it out

> The advantage of this approach is that only a single new [person](/docs/data/persons), and all group events that are submitted will be associated with this person.

Reviewer: Can you add an example of analysis you can do with this method?

<JSCreateGroups />

## How to set group properties

<JSSetGroupProperties />

## How to capture group events

Events are automatically linked to the groups that the user who submitted the event belongs to. Thus you do not add any additional code once you have linked a user to the group. This is the recommended way to capturing group events.

To link a user to a group, call the same code as you would when [creating a group](#how-to-create-groups).

### Advanced (server-side only): Capturing group events without a user

If you want to capture group events but don't want to associate them with a specific user, there are two alternatives:

> #### Method one (recommended): Use the group ID as the event's distinct ID

Reviewer 2: Why is this recommended?

I think last I discussed with @tiina303 , our recommendation was to create a proxy distinct ID, like "group_events" that is used for all group calls without a user, so we don't generate random useless people.

The confusion here might be that our SDKs don't follow this best practice, because earlier indeed we were doing the same 🙈

Author: Oh wait, I think I get it now: You can assign the event to a group, and it doesn't matter if the `distinct_id` of the user who submitted the event is assigning different groups to different events (I was under the impression that 1 user = 1 group. But I get it now that users and groups are not linked)

Reviewer 2: Exactly

<JSCreateGroups />

## How to set group properties

<JSSetGroupProperties />

> ## How to capture group events

Reviewer: ```suggestion
## How to capture non-user group events
```

<MultiLanguage>

> ```js-web

Reviewer: Lets just keep js-web stuff in one PR so we don't get confused across PRs.
